Queues

deyemiobaa opened the pull request:
This pull request contains the article for creating queues using JavaScript.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Codecademy/ugc?pullRequest=50) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Codecademy/ugc?pullRequest=50) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Codecademy/ugc?pullRequest=50) it.</sub>

KTom101 reviewed (changes requested):
@deyemiobaa Thank you for your entry. I've completed my review of this. I also noticed that the main file is not in markdown. When making the changes, can you convert the file into markdown so that it ends in the `.md` extension?   

Please let us know if you have any questions or when you're done making the changes. Thanks!

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
````diff
@@ -0,0 +1,154 @@
+  }
+}
+```
+
````
I also think this would be really helpful, great suggestion @KTom101!

deyemiobaa replied:
Great suggestion, working on that!

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+## Creating Queues Using JavaScript
+
+During conversations about data structures in the programming world, queues are heard quite often.
+A queue is a linear data structure that follows the first-in-first-out pattern i.e removal takes place at the front, and addition takes place at the end. Think of it as the checkout point of a grocery store. Customers who walk there first get attended to first.
```
I'd love to hear a little more about linear data structures and what it means to conform to this kind of structure. I think it would also be good to do a more detailed comparison on this grocery store checkout counter anecdote in regard to what the customer represents (a data point?) and how the customer that gets attended to then exits the queue.

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+
+There are several use cases for queues in the programming and real world. some them include:
+
+- A desk printer 
+- Adding songs to a queue in your music player
+- Customer service wait-lines
+- File sharing between / data transfer between two processes
```
Maybe we tack on a colon ":" to each of these list items and go into a little bit of detail about how these are related to queues!

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+- Customer service wait-lines
+- File sharing between / data transfer between two processes
+
+Queues can be implemented in any programming language but our focus is on how to do it using javaScript. In javascript, it can also be implemented in two ways (arrays and linked lists), for this article we would implement queues using arrays.
```
Added in an additional edit on top of @KTom101's

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+
+## Getting started
+
+First step is to initialize our class with it's own storage (an array where our queue items would be stores)
```
Another term we can use for item is "element." However, for consistency purposes, I think using the term item throughout this article is 👍

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
````diff
@@ -0,0 +1,154 @@
+
+This term refers to adding a new item to the queue. Since we're implementing our queue with an array, we can use the `push` array method to add new items to the queue.
+
+```js
+class Queue {
+  constructor() {
+    this.queue = [];
+  }
+
+  enqueue(item) {
+    this.queue.push(item);
+  }
+}
+```
````
After the code block, I think that it'd be helpful to break some of this syntax down. A short paragraph is sufficient and also a great opportunity to link some of the Docs entries on the topic, [JavaScript](https://www.codecademy.com/resources/docs/javascript).

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+
+### Dequeue
+
+This term refers to removing a new item from the queue. Again, the `shift` array method takes care of this for us easily.
```
For methods, (dw too much about this as it's a structural/formatting thing), we'll write them out like so: `.methodName()`. This is because unlike free functions, these are added onto a data type's prototype. For example, shift is called on an array prototype so it's written out as `.shift()` with the dot notation and parenthesis for invocation.

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
```diff
@@ -0,0 +1,154 @@
+For the implementation of the above operations, we’d use an [ES6 class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) and have the various operations as methods
+
+
+## Getting started
```
Could we change this heading to something more involved since we're also fitting the methods of a queue under this heading?

yangc95 commented on content/sodiq_aderibigbe/creating-queues-using-javascript:
````diff
@@ -0,0 +1,154 @@
+}
+```
+
+That’s it, folks. We’ve successfully implemented a queue and its basic operations using JavaScript.
````
Cool! Can we add a ## Conclusion heading here and a sentence or two more summing up everything in the article?

yangc95 reviewed (commented):
@deyemiobaa Great efforts and for putting your voice into this! Let's continue improving this article ☺️ Please let us know when you've looked over the suggestions/comments or have any questions on anything (even if it's something like asking for resources to learn more).

deyemiobaa commented:
Hi, @KTom101 @yangc95. Thank you for your review. I have made a couple of changes to the article using your suggestions as a guide. 
Please let me know if there's anything that still needs to be worked on.

KTom101 commented on content/sodiq_aderibigbe/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+
+The code block above represents the creation of a class, the [`constructor()`](https://www.codecademy.com/resources/docs/javascript/constructors) method is a special method of a class that used to create an object instance of that class. It can be used to initialize any object that would be used across the class methods. The [`this`](https://www.codecademy.com/resources/docs/javascript/this) keyw ...
+
+Next up, we add each queue operation as a method of the class
```
```suggestion
Next up, we add each queue operation as a method of the class.
```

KTom101 commented:
@deyemiobaa Thank you for making the changes. I've added a few more.

deyemiobaa commented:
@KTom101 I've made the highlighted changes.

Dusch4593 commented:
Hey @deyemiobaa 👋🏻 Brandon from the Free Team! Can we make sure that the parent folder for this article matches the user for your Codecademy profile? (is [this](https://www.codecademy.com/profiles/deyemiobaa) your username?)

deyemiobaa commented:
> > Hi @Dusch4593, that's my username.
> 
> @deyemiobaa Great! I just added an `author_meta.json` for your profile and updated the author folder to your Codecademy username.
> 
> Next time you're making any local changes in this PR branch, I would `git pull`, first.

Thanks, I have the updated folder now 👍🏾

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+
+  reverse() {
+    const reversed = []
+    while (this.queue.length > 0) {
```
Note: added spacing above this comment for readability.
```suggestion

    // iterate through the array using a while loop
    while (this.queue.length > 0) {
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+  reverse() {
+    const reversed = []
+    while (this.queue.length > 0) {
+      reversed.push(this.queue.pop());
```
```suggestion
      // remove the last element and push it to the new array
      reversed.push(this.queue.pop());
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+    while (this.queue.length > 0) {
+      reversed.push(this.queue.pop());
+    }
+    this.queue = reversed;
```
```suggestion
    // set queue using the new array
    this.queue = reversed;
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
````diff
@@ -0,0 +1,206 @@
+  }
+}
+```
+
````
I advise taking out lines 170-174. Place line 170 at the end of 138 and use the bullet points as comments within the snippet. I've provided examples of what this would look like above.

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+
+### Reversing a Queue
+
+As the title implies, we are simply trying to change the order of the queue from back to front.
```
```suggestion
As the title implies, we are simply trying to change the order of the queue from back to front. The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops).
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
````diff
@@ -0,0 +1,206 @@
+}
+```
+
+The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops). 
````
```suggestion
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
````diff
@@ -0,0 +1,206 @@
+```
+
+The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops). 
+- First we declare an empty array.
````
```suggestion
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+
+The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops). 
+- First we declare an empty array.
+- Go through each content in the queue `while (this.queue.length > 0)` till it's empty
```
```suggestion
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops). 
+- First we declare an empty array.
+- Go through each content in the queue `while (this.queue.length > 0)` till it's empty
+- Remove the last element and push it into the new array `reversed.push(this.queue.pop())`
```
```suggestion
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+- First we declare an empty array.
+- Go through each content in the queue `while (this.queue.length > 0)` till it's empty
+- Remove the last element and push it into the new array `reversed.push(this.queue.pop())`
+- Update the queue with the new array 
```
```suggestion
```

KTom101 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+- Go through each content in the queue `while (this.queue.length > 0)` till it's empty
+- Remove the last element and push it into the new array `reversed.push(this.queue.pop())`
+- Update the queue with the new array 
+
```
```suggestion
```

KTom101 commented:
@deyemiobaa I've added some more changes. This piece is looking very good. Please let me know if you have questions concerning any of the changes.

deyemiobaa commented:
Hi @KTom101. Thanks for the suggested changes, they were very helpful. I've made the changes to the article. Please let me know if there's anything else I need to adjust. Thank you.

KTom101 reviewed (approved):
@deyemiobaa Thanks for making the changes. I think this looks good to go. I'm going to tag @yangc95 to take a look at this to make sure they're happy with this, too.

deyemiobaa commented:
Thank you, @KTom101. The review process has been very helpful. I appreciate the time and effort you and @yangc95 put in to make the piece good.

yangc95 reviewed (approved):
@deyemiobaa I do apologize, I was making some changes with punctuation and grammar/structure but accidentally reversed your changes as a by-product of pushing. 

Rest assured, this article will be solely written by you. We've only made edits here to make sure everything is clear! Thanks again ☺️ this article looks good to me.

yangc95 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,206 @@
+result.enqueue(4) // Returns [5, 3, 4]
+result.enqueue(7) // Returns [5, 3, 4, 7]
+
+// removing an element
```
```suggestion
// Removing an element
```

Dusch4593 reviewed (changes requested):
Hey @deyemiobaa 👋🏻 I wanted to offer one more review of this article before we merge and publish. Let us know if you have any questions about any of the comments/suggestions. 😄

deyemiobaa commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,229 @@
+
+When transferring files from one device to another, the files are received on the other end in the same order they were sent, irrespective of the time it takes for any of them to get completed.
+
+## Implementation of Queues
+
+Queues can be implemented in any programming language, but our focus is on how to create them using JavaScript. In JavaScript, it can also be implemented in two ways: arrays and linked lists. For this article, we will implement queues using arrays.
+
```
@Dusch4593, I've implemented the suggested changes and added one to this part of the article. The paragraph was looking isolated so I added a header to it.

I also wanted to attach links to these two terms, `array` and `linked list`, but I could only find an article for arrays on the Codecademy blog so I left it. If you have any suggestions or you think it's fine this way, please let me know. Thanks

Dusch4593 replied:
That's fine! 😄 We can leave this as it is.

Dusch4593 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,229 @@
+
+### Dequeue
+
+This term refers to removing a new element from the queue. Again, the `.shift()` array method takes care of this for us easily.
```
```suggestion
This term refers to removing a new element from the queue. Again, the [`.shift()`](https://www.codecademy.com/resources/docs/javascript/arrays/shift) array method takes care of this for us easily.
```

Dusch4593 commented on content/deyemiobaa/creating-queues-using-javascript.md:
```diff
@@ -0,0 +1,229 @@
+
+### Reversing a Queue
+
+As the title implies, we are simply trying to change the order of the queue from back to front. The `reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops) and the [`.pop()`](https://www.codecademy.com/resources/docs/javascript/arrays/pop) array method.
```
```suggestion
As the title implies, we are simply trying to change the order of the queue from back to front. The `.reverse()` method is handled using a [while loop](https://www.codecademy.com/resources/docs/javascript/loops) and the [`.pop()`](https://www.codecademy.com/resources/docs/javascript/arrays/pop) array method.
```

Dusch4593 reviewed (changes requested):
@deyemiobaa Great work, thanks for pushing those changes 👍🏻 I left just three more comments/suggestions for where we could link to some more Docs entries.

deyemiobaa commented:
All done @Dusch4593!

Dusch4593 reviewed (approved):
@deyemiobaa Looks great! 👍🏻 I think this is ready to go, but I just wanted to let you know about some changes I just made in fb505861f7458884047fedbd10bd6ecbeb7db474 and faadddb8e7693fb555e7354bced0e2d0f418e369:

- The last snippet was unlabeled but was output. So I labeled it with \```shell.
- All snippets, headers, and lists should have an empty line on top of and underneath of them.
- Avoid using special characters like `-` and `/` except for `?`.
- Changed a few of the Lv. 3 headings to have less words.

Dusch4593 commented:
<img width="1149" alt="Screen Shot 2022-05-13 at 1 17 07 PM" src="https://user-images.githubusercontent.com/23144845/168335021-a49cfacb-e1ad-4bea-a90d-c1825542871b.png">

Congratulations @deyemiobaa! 🎉 👏🏻 🙌🏻 🎊 Your article is now published and live! You can check out here with this link: https://www.codecademy.com/article/deyemiobaa/creating-queues-using-javascript

Thank you very much for contributing this content and we look forward to working with you again! 😄

deyemiobaa commented:
Thank you @Dusch4593 😊🎉🙏🏾
